### PR TITLE
FBFW-121 - Get Messages API

### PR DESCRIPTION
**Pull request by mariomelchor**
Added API route to get messages.

Download PR. run `sh start-server.sh` Hit the following endpoints

Parameters allowed
gameID 
countryID 
toCountryID 
offset = default is 0
limit = default is 25

## All messages - This just requires to set countryID=0
http://localhost/api.php?route=game/getmessages&gameID=2&countryID=0

Screen Shot 2022-04-14 at 10 11 40 AM

## All messages w/ offset 
http://localhost/api.php?route=game/getmessages&gameID=2&countryID=0&offset=1

Screen Shot 2022-04-14 at 10 13 09 AM

## Limit = 1
http://localhost/api.php?route=game/getmessages&gameID=2&countryID=2&toCountryID=3&limit=1

Screen Shot 2022-04-14 at 10 29 44 AM

## Between Countries
http://localhost/api.php?route=game/getmessages&gameID=2&countryID=2&toCountryID=3

Screen Shot 2022-04-14 at 10 29 32 AM

## Accessing another countries messages - Returns no permissions messages

http://localhost/api.php?route=game/getmessages&gameID=2&countryID=3&toCountryID=2

Screen Shot 2022-04-19 at 12 39 48 PM (3)

## If there are no messages 

you get a successful request but an empty messages array

Screen Shot 2022-04-19 at 12 42 18 PM

## pressType == 'NoPress' 

Returns no messages

Screen Shot 2022-04-19 at 12 44 38 PM

## pressType == 'RulebookPress' during gamePhase == 'Builds' or gamePhase == 'Retreats'

Returns no messages

Screen Shot 2022-04-19 at 12 49 48 PM

## Not Logged In 
You get an NO API Key provided.

Screen Shot 2022-04-19 at 12 51 35 PM

### CODE REVIEW

File: api.php [hunk 1]
```diff
@@ -1008,6 +1008,61 @@ public function run($userID, $permissionIsExplicit) {
...
+		$args = $this->getArgs();
+
+		if ($args['gameID'] === null)
+			throw new RequestException('gameID is required.');
+
+		if ($args['countryID'] === null)
+			throw new RequestException('countryID is required.');
+
+		if (intval($args['limit']) > 25)
+			throw new RequestException('limit should not exceed 25');
```
Comment (bryancodazen) [inline api.php:1030]: You should use the JSONresponse builder here to send back JSON data in the event of an exception. 

Also, this should account for invalid gameID/countryID args such as the use of characters. You should probably do a numeric check here.

File: api.php [hunk 2]
```diff
@@ -1008,6 +1008,61 @@ public function run($userID, $permissionIsExplicit) {
...
+
+		$gameID = intval($args['gameID']);
+		$countryID = intval($args['countryID']);
+		$toCountryID = intval($args['toCountryID']);
+		$offset = intval($args['offset']);
+		$limit = intval($args['limit']);
+
+		$limit = $limit ? $limit : 25;
```
Comment (bryancodazen) [inline api.php:1038]: Let's make the `25` a variable since we're also using it in an exception handler above.

Comment (bryancodazen) [inline api.php:1038]: Let's make the `25` a constant since we're also using it in an exception handler above.

File: api.php [hunk 3]
```diff
@@ -1008,6 +1008,61 @@ public function run($userID, $permissionIsExplicit) {
...
+		$offset =  $offset ? $offset : 0;
```
Comment (bryancodazen) [inline api.php:1039]: can you have negative offsets?

Comment (bryancodazen) [inline api.php:1039]: can you have negative offsets?

File: api.php [hunk 4]
```diff
@@ -1008,6 +1008,61 @@ public function run($userID, $permissionIsExplicit) {
...
+		if ( $countryID == 0 ) {
+			$where = "toCountryID = 0";
+		}
+
+		// Only get messages sent between
+		else {
+			$where = "( toCountryID = $toCountryID AND fromCountryID = $countryID )
+					OR
+					( fromCountryID = $toCountryID AND toCountryID = $countryID )";
+		}
```
Comment (bryancodazen) [inline api.php:1051]: Is there any security risk here such as SQL injection?

File: api.php [hunk 5]
```diff
@@ -1008,6 +1008,61 @@ public function run($userID, $permissionIsExplicit) {
...
+
+		$tabl = $DB->sql_tabl("SELECT message, toCountryID, fromCountryID, turn, timeSent
+		FROM wD_GameMessages WHERE gameID = $gameID AND $where
+		order BY id DESC LIMIT $limit OFFSET $offset");
```
Comment (bryancodazen) [inline api.php:1055]: Is there any security risk here such as SQL injection?

Comment (bryancodazen) [inline api.php:1055]: Is there any security risk here such as SQL injection?

File: api.php [hunk 6]
```diff
@@ -1008,6 +1008,61 @@ public function run($userID, $permissionIsExplicit) {
...
+
+		$messages = array();
+		while ( $message = $DB->tabl_hash($tabl) ) {
+			$messages[] = $message;
+		}
+
+		return json_encode($messages);
```
Comment (bryancodazen) [inline api.php:1062]: use the JSON response builder

Comment (bryancodazen) [inline api.php:1062]: use the JSON response builder

File: api.php [hunk 7]
```diff
@@ -1271,6 +1326,7 @@ public function run() {
 	$api->load(new SetOrders());
 	$api->load(new ToggleVote());
 	$api->load(new SendMessage());
+	$api->load(new GetMessage());
```
Comment (bryancodazen) [inline api.php:1329]: should we call this `GetMessages` instead?

Comment (bryancodazen) [inline api.php:1329]: should we call this `GetMessages` instead?

**Review by bryancodazen — CHANGES_REQUESTED** [review]
I am able to see messages between other players by changing the route arguments. For example, in my local game there are three players:  Germany, England, and Austria. If I'm logged in as the Austria player, I can see the messages sent between the Germany/England players by changing the route args. 

There is session authentication being leveraged in `api.php`, I think you have to make use of this to make sure that players can only read messages they're supposed to. 

There are also 4 rules for press possible in a game which may affect what should be, should not be allowed when calling the API, have you factored this in?:
1. Regular
2. Public press only
3. No press
4. Rulebook press   


Screen Shot 2022-04-14 at 2 34 47 PM
Screen Shot 2022-04-14 at 2 34 59 PM
Screen Shot 2022-04-14 at 2 35 08 PM

File: api.php (hunk 3 above)
Comment (mariomelchor) [reply api.php:1039]: From what i can tell you cant have negative offsets.

https://stackoverflow.com/questions/17474507/negative-offset-in-mysql

**Comment by mariomelchor** [comment]
Still working on 

1. Only fetch messages for only the user’s messages
2. Based on game-mode/press type, limit API calls for messaging. 
3. Ensure you receive message if you are logged in.

File: api.php (hunk 4 above)
Comment (nebevets) [reply api.php:1058]: https://stackoverflow.com/questions/60174/how-can-i-prevent-sql-injection-in-php

Comment (mariomelchor) [reply api.php]: I looked into this database.php is already using mysqli. I also implemented very similar logic as is in chatbox.php

File: api.php [hunk 8]
```diff
@@ -1008,6 +1008,129 @@ public function run($userID, $permissionIsExplicit) {
...
+				'',
+				false,
+				[
+					'pressType' => $pressType,
+					'phase' => $gamePhase,
+				]
+			)
+		);
+		
+		if ($gameID === null || is_numeric($gameID) === false)
```
Comment (bryancodazen) [inline api.php:1053]: You can change `is_numeric === false` to `!is_numeric`

File: api.php [hunk 9]
```diff
@@ -1008,6 +1008,129 @@ public function run($userID, $permissionIsExplicit) {
...
+			throw new RequestException(
+				$this->JSONResponse('A gameID is required.', '', false, ['gameID' => $gameID])
+			);
+
+		if ($countryID === null || is_numeric($countryID) === false)
```
Comment (bryancodazen) [inline api.php:1058]: You can change `is_numeric === false` to `!is_numeric`

**Review by bryancodazen — CHANGES_REQUESTED** [review]
A work of art. Requesting some small changes.

**Review by justinyoocodazen — APPROVED** [review]
lgtm

**Review by bryancodazen — APPROVED** [review]
LGTM

**Pull request merged**